From dcff2201c35748df5fe9a8fe987cd8e73004a969 Mon Sep 17 00:00:00 2001
Message-Id: <dcff2201c35748df5fe9a8fe987cd8e73004a969.1414143271.git.chang-joon.lee@intel.com>
In-Reply-To: <f393e797ccaaf66a25ee3d7c4e6136083a6effeb.1414143271.git.chang-joon.lee@intel.com>
References: <f393e797ccaaf66a25ee3d7c4e6136083a6effeb.1414143271.git.chang-joon.lee@intel.com>
From: Gaurav K Singh <gaurav.k.singh@intel.com>
Date: Tue, 19 Aug 2014 00:30:34 +0530
Subject: [PATCH 17/26] FOR_UPSTREAM [VPG]: drm/i915: Fix the issue for Long
 packet send to DSI Controller

Fix the issue for Long packet send to DSI Controller. Make sure
that the generic data fifo is empty before sending long packet.
currently we are checking for fifo full. When the fifo full bit
is cleared by hardware there will be only space for 4 more bytes
to be pushed and hence it may not be enough to push long packet.

Issue: GMINL-1532
Change-Id: I37f4355b52167e38c6362afc84137ca8d1ddd539
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
Signed-off-by: Yogesh Mohan Marimuthu <yogesh.mohan.marimuthu@intel.com>
---
 drivers/gpu/drm/i915/intel_dsi_cmd.c |   19 +++++++++++++------
 1 file changed, 13 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_dsi_cmd.c b/drivers/gpu/drm/i915/intel_dsi_cmd.c
index ed39868..abbda65 100644
--- a/drivers/gpu/drm/i915/intel_dsi_cmd.c
+++ b/drivers/gpu/drm/i915/intel_dsi_cmd.c
@@ -172,7 +172,7 @@ static int dsi_vc_send_long(struct intel_dsi *intel_dsi, int channel,
 	struct drm_device *dev = encoder->dev;
 	struct drm_i915_private *dev_priv = dev->dev_private;
 	enum pipe pipe = intel_dsi->port;
-	u32 data_reg;
+	u32 data_reg, ctrl_reg, ctrl;
 	int i, j, n;
 	u32 mask;
 
@@ -181,14 +181,16 @@ static int dsi_vc_send_long(struct intel_dsi *intel_dsi, int channel,
 
 	if (intel_dsi->hs) {
 		data_reg = MIPI_HS_GEN_DATA(pipe);
-		mask = HS_DATA_FIFO_FULL;
+		ctrl_reg = MIPI_HS_GEN_CTRL(pipe);
+		mask = HS_CTRL_FIFO_EMPTY | HS_DATA_FIFO_EMPTY;
 	} else {
 		data_reg = MIPI_LP_GEN_DATA(pipe);
-		mask = LP_DATA_FIFO_FULL;
+		ctrl_reg = MIPI_LP_GEN_CTRL(pipe);
+		mask = LP_CTRL_FIFO_EMPTY | LP_DATA_FIFO_EMPTY;
 	}
 
-	if (wait_for((I915_READ(MIPI_GEN_FIFO_STAT(pipe)) & mask) == 0, 50))
-		DRM_ERROR("Timeout waiting for HS/LP DATA FIFO !full\n");
+	if (wait_for((I915_READ(MIPI_GEN_FIFO_STAT(pipe)) & mask) == mask, 50))
+		DRM_ERROR("Timeout waiting for HS/LP DATA FIFO to be empty\n");
 
 	for (i = 0; i < len; i += n) {
 		u32 val = 0;
@@ -201,8 +203,13 @@ static int dsi_vc_send_long(struct intel_dsi *intel_dsi, int channel,
 		/* XXX: check for data fifo full, once that is set, write 4
 		 * dwords, then wait for not set, then continue. */
 	}
+	ctrl = len << LONG_PACKET_WORD_COUNT_SHIFT;
+	ctrl |= channel << VIRTUAL_CHANNEL_SHIFT;
+	ctrl |= data_type << DATA_TYPE_SHIFT;
+
+	I915_WRITE(ctrl_reg, ctrl);
+	return 0;
 
-	return dsi_vc_send_short(intel_dsi, channel, data_type, len);
 }
 
 static int dsi_vc_write_common(struct intel_dsi *intel_dsi,
-- 
1.7.9.5

